File: sections/100_grandmaster_finale/problems/04_historical_rectangle_selection/solution.py
import sys

MOD = 1_000_000_007


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    n = data[0]
    x = [0] * n
    y = [0] * n
    reward = [0] * n
    at = 1
    for i in range(n):
        x[i], y[i], reward[i] = data[at:at + 3]
        at += 3
    ys = sorted(set(y))
    rank_y = [0] * n
    position = {value: i + 1 for i, value in enumerate(ys)}
    for i in range(n):
        rank_y[i] = position[y[i]]

    dp = reward.copy()
    ways = [1] * n
    bit_score = [0] * (len(ys) + 1)
    bit_ways = [0] * (len(ys) + 1)
    touched = []

    def update(pos: int, score: int, count: int) -> None:
        while pos < len(bit_score):
            if score > bit_score[pos]:
                bit_score[pos] = score
                bit_ways[pos] = count
            elif score == bit_score[pos] and score:
                bit_ways[pos] = (bit_ways[pos] + count) % MOD
            touched.append(pos)
            pos += pos & -pos

    def query(pos: int) -> tuple[int, int]:
        score = count = 0
        while pos:
            if bit_score[pos] > score:
                score, count = bit_score[pos], bit_ways[pos]
            elif bit_score[pos] == score and score:
                count = (count + bit_ways[pos]) % MOD
            pos -= pos & -pos
        return score, count

    sys.setrecursionlimit(1_000_000)

    def cdq(left: int, right: int) -> None:
        if right - left == 1:
            return
        middle = (left + right) // 2
        cdq(left, middle)
        first = sorted(range(left, middle), key=lambda i: (x[i], y[i]))
        second = sorted(range(middle, right), key=lambda i: (x[i], y[i]))
        pointer = 0
        for j in second:
            while pointer < len(first) and x[first[pointer]] < x[j]:
                i = first[pointer]
                pointer += 1
                update(rank_y[i], dp[i], ways[i])
            best, count = query(rank_y[j] - 1)
            if best:
                candidate = best + reward[j]
                if candidate > dp[j]:
                    dp[j], ways[j] = candidate, count
                elif candidate == dp[j]:
                    ways[j] = (ways[j] + count) % MOD
        for pos in touched:
            bit_score[pos] = bit_ways[pos] = 0
        touched.clear()
        cdq(middle, right)

    cdq(0, n)
    best = max(dp)
    count = sum(ways[i] for i in range(n) if dp[i] == best) % MOD
    print(best, count)


if __name__ == "__main__":
    main()

File: sections/100_grandmaster_finale/problems/04_historical_rectangle_selection/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using int64 = long long;

constexpr int MOD = 1'000'000'007;

struct State {
    int64 score = 0;
    int ways = 0;
};

State combine(State a, State b) {
    if (a.score != b.score) return a.score > b.score ? a : b;
    if (a.score == 0) return {0, 0};
    return {a.score, static_cast<int>((a.ways + static_cast<long long>(b.ways)) % MOD)};
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;
    vector<int> x(n), y(n), compressed_y;
    vector<int64> reward(n), dp(n);
    vector<int> ways(n, 1);
    for (int i = 0; i < n; ++i) {
        cin >> x[i] >> y[i] >> reward[i];
        compressed_y.push_back(y[i]);
        dp[i] = reward[i];
    }
    sort(compressed_y.begin(), compressed_y.end());
    compressed_y.erase(unique(compressed_y.begin(), compressed_y.end()),
                       compressed_y.end());
    vector<int> rank_y(n);
    for (int i = 0; i < n; ++i)
        rank_y[i] = lower_bound(compressed_y.begin(), compressed_y.end(), y[i])
                  - compressed_y.begin() + 1;

    vector<State> bit(compressed_y.size() + 1);
    vector<int> touched;
    auto update = [&](int position, State state) {
        for (int i = position; i < static_cast<int>(bit.size()); i += i & -i) {
            bit[i] = combine(bit[i], state);
            touched.push_back(i);
        }
    };
    auto query = [&](int position) {
        State result;
        for (int i = position; i > 0; i -= i & -i)
            result = combine(result, bit[i]);
        return result;
    };

    auto cdq = [&](auto &&self, int left, int right) -> void {
        if (right - left == 1) return;
        int middle = (left + right) / 2;
        self(self, left, middle);
        vector<int> first, second;
        first.reserve(middle - left);
        second.reserve(right - middle);
        for (int i = left; i < middle; ++i) first.push_back(i);
        for (int i = middle; i < right; ++i) second.push_back(i);
        sort(first.begin(), first.end(), [&](int a, int b) {
            return pair{x[a], y[a]} < pair{x[b], y[b]};
        });
        sort(second.begin(), second.end(), [&](int a, int b) {
            return pair{x[a], y[a]} < pair{x[b], y[b]};
        });
        size_t pointer = 0;
        for (int j : second) {
            while (pointer < first.size() && x[first[pointer]] < x[j]) {
                int i = first[pointer++];
                update(rank_y[i], {dp[i], ways[i]});
            }
            State predecessor = query(rank_y[j] - 1);
            if (predecessor.score == 0) continue;
            int64 candidate = predecessor.score + reward[j];
            if (candidate > dp[j]) {
                dp[j] = candidate;
                ways[j] = predecessor.ways;
            } else if (candidate == dp[j]) {
                ways[j] = (ways[j] + static_cast<long long>(predecessor.ways)) % MOD;
            }
        }
        for (int position : touched) bit[position] = {};
        touched.clear();
        self(self, middle, right);
    };
    cdq(cdq, 0, n);

    State answer;
    for (int i = 0; i < n; ++i)
        answer = combine(answer, {dp[i], ways[i]});
    cout << answer.score << ' ' << answer.ways << '\n';
}
